src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -693,6 +694,16 @@ private void RegisterApplicationLifetimeEvents()
             _applicationLifetime.ApplicationStopping.Register(() =>
             {
                 Interlocked.Exchange(ref _applicationStopping, 1);
+                if (_environment.IsDrainOnApplicationStopping())
+                {
+                    var drainModeManager = _host?.Services.GetService<IDrainModeManager>();
+                    if (drainModeManager != null)
+                    {
+                        _logger.LogInformation("Application Stopping: enabling drain mode");
+                        drainModeManager.EnableDrainModeAsync(CancellationToken.None).Wait();
```

Comment:
Is wait needed? If not, use similar approach to: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Controllers/HostController.cs#L115

Also, if wait is needed - pass in timeout

Reply:
Yes, a synchronous wait is needed here, because otherwise .NET assumes that the process is ready to shutdown. I'm not sure what timeout to set since the original ask was to enable a function to run for 8 or 10 hours to completion. If we add a default timeout, then I think it'll need to be configurable somewhere to allow those scenarios. Should that go in host.json?

Reply:
btw, in Kubernetes that timeout is managed externally. Kubernetes will sent the process a SIGTERM, which invokes `ApplicationStopping`, then waits for a set time (default 30 seconds, but configurable) then it'll send a SIGKILL, which will immediately invoke `ApplicationStopped` regardless of the wait here.

Reply:
Tagging @divyagandhisethi - as Kubernetes scenario is similar to how drain mode is handled in Linux Consumption -  `Drain` api is invoked on the host which puts instance in `DrainMode` and then after a pre-configured timeout - `SIGKILL` is sent which terminates the container.

Can we take similar approach here and  call `Drain` API?

Reply:
Yes, interested to know how we will determine the timeout between SIGTERM and SIGKILL. I think the same solution as https://github.com/Azure/azure-functions-host/issues/7188 is needed here - an endpoint on the host which tells us when all function invocations have been drained and it is safe to kill the instance at that point.

Reply:
@ahmelsayed - In Kubernetes -  Is it possible to explicitly call `Drain`  api on `SIGTERM` ? If yes, we can focus on looking into design for #7188

Reply:
Yes, that's what this PR is doing. `SIGTERM` on the process causes .NET to fire `ApplicationStopping`, which in this PR I call drainModeManager.EnableDrainModeAsync()

Reply:
> Yes, that's what this PR is doing.

Sorry - I should have been  more explicit - can we call `Drain` API outside of the host - similar to other skus - Platform makes a call  to Drain API

Reply:
oh sorry I misunderstood. We could add a `preStop` hook on the container that looks like this 

```yaml
lifecycle:
  preStop:
    exec:
      command: ["/bin/sh", "-c", "command to drain the host"]
# or
lifecycle:
  preStop:
    httpGet:
      path: "/admin/host/..."
```
This will execute a command, or sent a request to the container right before it sends SIGTERM. 

There are a couple of downsides to this:
1 - every functions user on kubernetes will need to know how to do this specifically for function deployment. We can auto do it for managed hostings.
2- Auth on the API. We could have a command line utility in the functions container that knows how to call that API on the host.

Reply:
Is this one way transaction? Is there a chance this container would be reused / activated. Also how would it impact other hosting platform App service plan, windows consumption. My main concern is that with indefinite wait it might create orphan hosts. Also we have work in progress for a resume API, So we can quickly start processing events incase there is an inflow of new events. In that case even though drain mode is initiated the same w3wp process might be used to quickly instantiate a new host and this change could potentially create orphan hosts.

Reply:
Thanks! Let's do the call in the host then.

Reply:
Another question: if container will not be terminated until sigkill is called, do we still need synchronous wait?

Reply:
> Is this one way transaction? Is there a chance this container would be reused / activated.

In kubernetes (or .NET core on linux in general) it is one way. A process/container that recieves SIGTERM is expected to be terminated. It'll not be reused. 

This might affect the behavior on Windows, hence it's only enabled by default for Kubernetes and with a flag. 

> Another question: if container will not be terminated until sigkill is called, do we still need synchronous wait?

Yes, sigkill is a last option if the application doesn't terminate on sigterm. The application is expected to terminate on sigterm, but it can trap the signal perform clean up/graceful shutdown. it can't stop sigkill as that will force the process to exit from the OS.